hal/halx86/acpi/madt.c

```diff
@@ -12,77 +13,113 @@
-            LocalApic->Header.Length == sizeof(ACPI_MADT_LOCAL_APIC))
+        switch (AcpiHeader->Type)
         {
-            ValidProcessorCount = HalpApicInfoTable.ProcessorCount;
-
-            HalpProcessorIdentity[ValidProcessorCount].LapicId = LocalApic->Id;
-            HalpProcessorIdentity[ValidProcessorCount].ProcessorId = LocalApic->ProcessorId;
-
-            HalpApicInfoTable.ProcessorCount++;
-
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
-        }
-        else
-        {
-            /* End the parsing early if we don't use the currently selected table */
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + 1);
+            case ACPI_MADT_TYPE_LOCAL_APIC:
+            {
+                ACPI_MADT_LOCAL_APIC *LA = (ACPI_MADT_LOCAL_APIC *)AcpiHeader;
+
+                DPRINT(" Local Apic, Processor %lu: ProcessorId %u, Id %u, LapicFlags %08X\n",
+                       HalpApicInfoTable.ProcessorCount,
+                       LA->ProcessorId, LA->Id, LA->LapicFlags);
+
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorId = LA->ProcessorId;
+                HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].LapicId = LA->Id;
+#if 0 // TODO: Is that correct?
+                if (LA->LapicFlags & 0x00000001) // Enabled
+                    HalpProcessorIdentity[HalpApicInfoTable.ProcessorCount].ProcessorStarted = TRUE;
+#endif
+
+                HalpApicInfoTable.ProcessorCount++;
+
+                break;
+            }
+            case ACPI_MADT_TYPE_IO_APIC:
+            {
+#if DBG && !defined(NDEBUG)
+                ACPI_MADT_IO_APIC *IA = (ACPI_MADT_IO_APIC *)AcpiHeader;
+
+                DPRINT(" Io Apic: Id %u, Address %08X, GlobalIrqBase %08X / UNIMPLEMENTED\n",
+                       IA->Id, IA->Address, IA->GlobalIrqBase);
+#endif
+
+                // TODO: Implement it.
+
+                break;
+            }
+            case ACPI_MADT_TYPE_INTERRUPT_OVERRIDE:
+            {
+#if DBG && !defined(NDEBUG)
+                ACPI_MADT_INTERRUPT_OVERRIDE *IO = (ACPI_MADT_INTERRUPT_OVERRIDE *)AcpiHeader;
+
+                DPRINT(" Interrupt Override: Bus %u, SourceIrq %u, GlobalIrq %08X, IntiFlags %04X / UNIMPLEMENTED\n",
+                       IO->Bus, IO->SourceIrq, IO->GlobalIrq, IO->IntiFlags);
+#endif
+
+                // TODO: Implement it.
+
+                break;
+            }
+            default:
+            {
+                DPRINT1(" UNIMPLEMENTED: Type %u, Length %u\n", AcpiHeader->Type, AcpiHeader->Length);
+            }
         }
+
+        AcpiHeader = (ACPI_SUBTABLE_HEADER *)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
     }
 }
 
 VOID
 HalpPrintApicTables(VOID)
-{ 
-    UINT32 i;
+{
+    ULONG i;
 
-    DPRINT1("HAL has detected a physical processor count of: %d\n", HalpApicInfoTable.ProcessorCount);
+    DPRINT1("Physical processor count: %lu\n", HalpApicInfoTable.ProcessorCount);
```

Comment:
Rewriting the other function makes sense. What's the point of rewriting this one?

Reply:
Initially, reduce number of printed lines and text lengths, also use consistent string formats.
Now, also print all data.